Clockwork/Support/Vanilla/Clockwork.php

```diff
@@ -456,14 +461,29 @@ protected function response($data = null, $status = null, $json = true)
 	// Make a Clockwork incoming request instance
 	protected function incomingRequest()
 	{
-		return new IncomingRequest([
-			'method'  => $_SERVER['REQUEST_METHOD'],
-			'uri'     => $_SERVER['REQUEST_URI'],
-			'input'   => $_REQUEST,
-			'cookies' => $_COOKIE
+		if ($this->incomingRequest) return $this->incomingRequest;
+
+		return $this->incomingRequest = new IncomingRequest([
+			'method'  => $this->psrRequest ? $this->psrRequest->getMethod() : $_SERVER['REQUEST_METHOD'],
+			'uri'     => $this->psrRequest ? $this->psrRequest->getUri()->getPath() : $_SERVER['REQUEST_URI'],
+			'headers' => $this->psrRequest
+				? array_map(function ($values) { return implode(', ', $values); }, $this->psrRequest->getHeaders())
+				: $_SERVER,
+			'input'   => $this->psrRequest
+				? array_merge($this->psrRequest->getQueryParams(), (array) $this->psrRequest->getParsedBody())
+				: array_merge($_GET, $_POST, (array) json_decode(file_get_contents('php://input'), true))
```

Comment:
I'm not sure whether the `'input'` has the same content with the two implementations. 

That said, I believe this could become a bit easier to read if it checked `$this->psrRequest` only once:

```php
return $this->incomingRequest = new IncomingRequest(
    $this->psrRequest ? [
        'method' => $this->psrRequest->getMethod(),
        'uri'    => $this->psrRequest->getUri()->getPath(),
        ..
    ] : [
        'method' => $_SERVER['REQUEST_METHOD'],
        'uri'    => $_SERVER['REQUEST_URI'],
        ..
    ]
);

Reply:
Fixed, now it should work even for metadata update requests with JSON bodies. Also split into separate methods to look more clean.